documents/Secure_Tunnel_Userguide.md

```diff
@@ -0,0 +1,165 @@
+
+```
+// Create Secure Tunnel Builder
+SecureTunnelBuilder builder = SecureTunnelBuilder(...);
+
+// Setting the onMessageReceived callback using the builder
+builder.WithOnMessageReceived([&](SecureTunnel *secureTunnel, const MessageReceivedEventData &eventData) {
+        {
+            std::shared_ptr<Message> message = eventData.message;
+            if (message->getServiceId().has_value()){
+                fprintf(
+                    stdout,
+                    "Message received on service id:'" PRInSTR "'\n",
+                    AWS_BYTE_CURSOR_PRI(message->getServiceId().value()));
+            }
+
+            if(message->getPayload().has_value()){
+                fprintf(
+                    stdout,
+                    "Message has payload:'" PRInSTR "'\n",
+                    AWS_BYTE_CURSOR_PRI(message->getPayload().value()));
+            }
+        }
+    });
+
+// Build Secure Tunnel Client
+std::shared_ptr<SecureTunnel> secureTunnel = builder.Build();
+
+if (secureTunnel == nullptr)
+{
+    fprintf(stdout, "Secure Tunnel creation failed.\n");
+    return -1;
+}
+
+// Start the secure tunnel connection
+if (!secureTunnel->Start())
+{
+    fprintf("Failed to start Secure Tunnel\n");
+    return -1;
+}
+
+// Messages received on a stream will now be printed to stdout.
+```
+
+# How to Start and Stop
+
+## Start
+Invoking Start() on the Secure Tunnel Client will put it into an active state where it recurrently establishes a connection to the configured Secure Tunnel endpoint using the provided [Client Access Token](https://docs.aws.amazon.com/iot/latest/developerguide/secure-tunneling-concepts.html). If a [Client Token](https://docs.aws.amazon.com/iot/latest/developerguide/secure-tunneling-concepts.html) is provided, the Secure Tunnel Client will use it but in the event one is not provided, the Secure Tunnel Client will automatically generate one for use on a reconnection attempts. The Client Token for any initial connection to the Secure Tunnel service MUST be unique. Reusing a Client Token from a previous connection will result in a failed connection to the Secure Tunnel Service.
+```
+// Create Secure Tunnel Builder
+SecureTunnelBuilder builder = SecureTunnelBuilder(...);
+
+// Adding a client token to the builder
+String clientToken;
+builder.WithClientToken(clientToken.c_str());
+
+// Build Secure Tunnel Client
+std::shared_ptr<SecureTunnel> secureTunnel = builder.Build();
+
+if (secureTunnel == nullptr)
+{
+    fprintf(stdout, "Secure Tunnel creation failed.\n");
+    return -1;
+}
+
+// Start the secure tunnel connection
+if (!secureTunnel->Start())
+{
+    fprintf("Failed to start Secure Tunnel\n");
+    return -1;
+}
+```
+
+## Stop
+Invoking Stop() on the Secure Tunnel Client breaks the current connection (if any) and moves the client into an idle state.
```

Comment:
I might replace `(if any)` to simply `if one exists`.

Minor: Put `Stop()` in a code block, add `cpp` to the big code block below.